Duplicated figures with BackgroundPlotter in sphinx

GuillaumeFavelier opened the pull request:
I have an issue of duplicated figures with `BackgroundPlotter` and `sphinx-gallery`. I share my example in here.

I expect one figure after cell 1 but we can see two and I expect no figure in cell 2 since there is a call to `close()` but once again we can see two.

![image](https://user-images.githubusercontent.com/18143289/72742785-bc6d9a00-3baa-11ea-9e8d-cace9fd1f1ba.png)

banesullivan commented:
Hm. I’ll look into this since I wrote the image scraper for Sphinx. I didn’t really think about the BackgroundPlotter when making it, so likely a quick bug fix

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

banesullivan commented:
So the issue is arising because the `BackgroundPlotter` is added to the `_ALL_PLOTTERS` list twice... this is concerning. As the only place any plotter is added is in the `BasePlotter.__init__` method. So this means that the super call is happening twice?

https://github.com/pyvista/pyvista/blob/fce5954590a317784155324aee9df04427af7ff9/pyvista/plotting/plotting.py#L185-L187

banesullivan commented:
From what I can tell, the following call also calls the `BasePlotter.__init__` which deosn't make sense to me:

https://github.com/pyvista/pyvista/blob/fce5954590a317784155324aee9df04427af7ff9/pyvista/plotting/qt_plotting.py#L364

Found this out by raising a `RuntimeError` in `BasePlotter.__init__`

<details>

```py
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-43985e843322> in <module>
      1 import pyvista as pv
      2 
----> 3 p = pv.BackgroundPlotter()
      4 p.close()
      5 pv.plotting._ALL_PLOTTERS

~/Software/pyvista/pyvista/pyvista/plotting/qt_plotting.py in __init__(self, show, app, shape, window_size, off_screen, auto_update, **kwargs)
    583         QtInteractor.__init__(self, parent=self.frame, shape=shape,
    584                               off_screen=off_screen, auto_update=auto_update,
--> 585                               **kwargs)
    586         self.app_window.signal_close.connect(self.quit)
    587         self.add_toolbars(self.app_window)

~/Software/pyvista/pyvista/pyvista/plotting/qt_plotting.py in __init__(self, parent, title, shape, off_screen, border, border_color, border_width, multi_samples, line_smoothing, point_smoothing, polygon_smoothing, splitting_position, auto_update)
    362         if not has_pyqt:
    363             raise AssertionError('Requires PyQt5')
--> 364         QVTKRenderWindowInteractor.__init__(self, parent)
    365         BasePlotter.__init__(self, shape=shape, title=title,
    366                              border=border, border_color=border_color,

~/anaconda3/envs/dev/lib/python3.7/site-packages/vtkmodules/qt/QVTKRenderWindowInteractor.py in __init__(self, parent, **kw)
    255             QWidget.__init__(self, parent, wflags | Qt.MSWindowsOwnDC)
    256         elif QVTKRWIBase == "QGLWidget":
--> 257             QGLWidget.__init__(self, parent)
    258 
    259         if rw: # user-supplied render window

~/Software/pyvista/pyvista/pyvista/plotting/plotting.py in __init__(self, shape, border, border_color, border_width, title, splitting_position)
    185         # Add self to open plotters
    186         self._id_name = "{}-{}".format(str(hex(id(self))), len(_ALL_PLOTTERS))
--> 187         raise RuntimeError()
    188         _ALL_PLOTTERS[self._id_name] = self
    189 

RuntimeError: 
```

</details>

GuillaumeFavelier commented:
Thanks for diving into this @banesullivan!

It is maybe because `QtInteractor` inherits from both `QVTKRenderWindow` and `BasePlotter`.

This situation probably didn't happen before because the way we were pushing the plotters was  using `id()`.

banesullivan commented:
> This situation probably didn't happen before because the way we were pushing the plotters into _ALL_PLOTTERS was using id().

Exactly right. So now that we are always using a unique string to keep track of instantiated plotters, we see `BackgroundPlotter`s call that `__init__` twice. This means renderers and all that are being set up twice, which probably isn't good as they the first ones might not get cleaned up properly?

I'm not sure how to fix this at the moment...

GuillaumeFavelier commented:
Seems related to: https://stackoverflow.com/questions/9575409/calling-parent-class-init-with-multiple-inheritance-whats-the-right-way/9575426#9575426

GuillaumeFavelier commented:
Removing the explicit call to `BasePlotter.__init__()` in `QtInteractor.__init__()` seems to fix the duplicate call for me. I need to check that all parameters are passed properly though. But the figure is still there after cell 2, is this behaviour expected? Maybe I missd something.

![image](https://user-images.githubusercontent.com/18143289/72797443-6ef03c00-3c41-11ea-87f2-6411c6474ffc.png)

I'll investigate further.

GuillaumeFavelier commented:
This was not a good idea, I checked and the parameters are not passed correctly.

GuillaumeFavelier commented:
Keeping the stackoverflow thread in mind, I tried to reproduce a simplified version of our design. I think that we are close to the following:

```py
from PyQt5.QtOpenGL import QGLWidget
from PyQt5.QtWidgets import QApplication

app = QApplication.instance()
app = QApplication([''])


class QVTKRenderWindowInteractor(QGLWidget):
    def __init__(self):
        print("Entering QVTKRenderWindowInteractor")
        QGLWidget.__init__(self)
        print("Leaving QVTKRenderWindowInteractor")


class BasePlotter(object):
    def __new__(cls):
        if cls is BasePlotter:
            raise TypeError("BasePlotter is an abstract class.")
        return object.__new__(cls)

    def __init__(self):
        print("Entering BasePlotter")
        print("Leaving BasePlotter")


class QtInteractor(QVTKRenderWindowInteractor, BasePlotter):
    def __init__(self):
        print("Entering QtInteractor")
        QVTKRenderWindowInteractor.__init__(self)
        BasePlotter.__init__(self)
        print("Leaving QtInteractor")


class BackgroundPlotter(QtInteractor):
    def __init__(self):
        print("Entering BackgroundPlotter")
        QtInteractor.__init__(self)
        print("Leaving BackgroundPlotter")


p = BackgroundPlotter()
```

which gives:

```
Entering BackgroundPlotter
Entering QtInteractor
Entering QVTKRenderWindowInteractor
Entering BasePlotter
Leaving BasePlotter
Leaving QVTKRenderWindowInteractor
Entering BasePlotter
Leaving BasePlotter
Leaving QtInteractor
Leaving BackgroundPlotter
```

GuillaumeFavelier commented:
This leads to duplicated calls to `BasePlotter.__init__()`. The first one with default parameters and the second with the correct parameters. This explains why the values were not passed after removing the explicit call.

A short term solution to mitigate would still be to remove the second call and set the parameters afterwards. I'm not sure if it's viable.

GuillaumeFavelier commented:
Another solution would be to design an adapter class:

```py
...

class QVTKRenderWindowInteractorAdapter(object):
    def __init__(self, **kwargs):
        self.interactor = QVTKRenderWindowInteractor()
        super(QVTKRenderWindowInteractorAdapter, self).__init__(**kwargs)


class QtInteractor(QVTKRenderWindowInteractorAdapter, BasePlotter):
    def __init__(self, **kwargs):
        print("Entering QtInteractor")
        super(QtInteractor, self).__init__(**kwargs)
        print("Leaving QtInteractor")


class BackgroundPlotter(QtInteractor):
    def __init__(self, **kwargs):
        print("Entering BackgroundPlotter")
        super(BackgroundPlotter, self).__init__(**kwargs)
        print("Leaving BackgroundPlotter")


p = BackgroundPlotter(param=1)
```

which gives:

```
Entering BackgroundPlotter
Entering QtInteractor
Entering QVTKRenderWindowInteractor
Leaving QVTKRenderWindowInteractor
Entering BasePlotter
1
Leaving BasePlotter
Leaving QtInteractor
Leaving BackgroundPlotter
```

There is only one call with the correct value.

GuillaumeFavelier commented:
This PR introduces `QVTKRenderWindowAdapter` that relies on `super()` to resolve the original inheritance issue and avoid the duplicated call to `BasePlotter.__init__()`. It also unifies the render calls of either `Plotter` or `BackgroundPlotter` through a unique `render()` function now. This change requires extensive testing to avoid regression. This is still a work in progress.

GuillaumeFavelier commented:
I think that the `BackgroundPlotter` should be able to select the correct version of the methods inherited from `BasePlotter`. So if you don't mind I can continue refactoring in here especially the signal/slot system which I think are not all necessary. For example, `set_background()` has trouble with the `top` parameter.

GuillaumeFavelier commented:
Results of testing on `mne-python` are available: 
https://17639-1301584-gh.circle-artifacts.com/0/dev/index.html

Original issue was reported because of this example: 
https://17639-1301584-gh.circle-artifacts.com/0/dev/auto_examples/visualization/plot_publication_figure.html#sphx-glr-auto-examples-visualization-plot-publication-figure-py

Good news is that there is no more duplicated figure, I will continue to inspect more examples to detect anomalies. But a figure is still displayed after a call to `close()`

GuillaumeFavelier commented:
Maybe it could be helpful for the pyvista scraper to differentiate if a plotter is closed because of the user `close()` or because of `pyvista.close_all()`

GuillaumeFavelier commented:
This is already a lot of changes to tackle duplicated figures so I would be okay to take care of the closing issue somewhere else.

GuillaumeFavelier commented:
Next idea is a test commit that builds (close to?) all PyVista documentation examples with BackgroundPlotter.

banesullivan commented on pyvista/plotting/qt_plotting.py:
```diff
@@ -432,7 +432,7 @@ def __init__(self, parent=None, title=None, off_screen=None,
 
         if rcParams["depth_peeling"]["enabled"]:
             for renderer in self.renderers:
-                self.enable_depth_peeling()
```
Good catch. Can you make sure to update this branch with `master`. Also, note that depth peeling needs `self.ren_win.AlphaBitPlanesOn()`... is this done here and for the `Plotter` class?

`AlphaBitPlanesOn` was forgotten in #529 and might have something to do with the issues you had in https://github.com/pyvista/pyvista/pull/529#issuecomment-574076243

banesullivan replied:
same needs to be done for `Plotter`

GuillaumeFavelier replied:
I guess we you could make a call from `self` first and depending on the `result` iterate through the renderers? What do you think?

GuillaumeFavelier replied:
Or port all this in `BasePlotter` actually since it would be the same for `Plotter`

banesullivan commented on pyvista/plotting/qt_plotting.py:
```diff
@@ -537,6 +537,11 @@ def remove_actor(self, actor, reset_camera=None):
+    def render(self):
+        """Render the main window."""
+        if hasattr(self, 'ren_win'):
+            self.ren_win.Render()
```
is this overriding and inherited method? Currently, if I call `render()`, an error from Qt is thrown

GuillaumeFavelier replied:
No it's not supposed to be overriding. It's the symmetric of `Plotter`'s `render`. I don't know why you have an error here. 
You call `render` through `BackgroundPlotter`?

GuillaumeFavelier commented:
The latest commit is just a test building the documentation (almost) entirely with `BackgroundPlotter`.
Locally, I had issues with the following examples:
* examples/02-plot/orbit.py
* examples/02-plot/topo-map.py

GuillaumeFavelier commented:
I was a bit aggressive on the testing by using `Plotter=BackgroundPlotter`. I can try by replacing directly in the example.

GuillaumeFavelier commented:
In this test commit, the examples of the documentation using explicitly `Plotter` use now `BackgroundPlotter`:

```
examples/00-load/create-geometric-objects.py
examples/00-load/create-point-cloud.py
examples/00-load/read-file.py
examples/01-filter/clipping.py
examples/01-filter/connectivity.py
examples/01-filter/decimate.py
examples/01-filter/extract-cells-inside-surface.py
examples/01-filter/extract-edges.py
examples/01-filter/gaussian-smoothing.py
examples/01-filter/geodesic.py
examples/01-filter/glyphs.py
examples/01-filter/poly-ray-trace.py
examples/01-filter/project-plane.py
examples/01-filter/resample.py
examples/01-filter/slicing.py
examples/01-filter/streamlines.py
examples/01-filter/using-filters.py
examples/02-plot/bgplot.py
examples/02-plot/clear.py
examples/02-plot/cmap.py
examples/02-plot/depth-peeling.py
examples/02-plot/edl.py
examples/02-plot/gif.py
examples/02-plot/image_depth.py
examples/02-plot/interpolate-before-map.py
examples/02-plot/labels.py
examples/02-plot/lighting.py
examples/02-plot/linked.py
examples/02-plot/movie.py
examples/02-plot/multi-window.py
examples/02-plot/opacity.py
examples/02-plot/plot-over-line.py
examples/02-plot/scalar-bars.py
examples/02-plot/screenshot.py
examples/02-plot/spherical.py
examples/02-plot/themes.py
examples/02-plot/volume.py
```

Those examples were not moved to `BackgroundPlotter` because of a segfault:

```
examples/02-plot/orbit.py
examples/02-plot/topo-map.py
```

Implicit call to `Plotter` through the `plot()` function were not moved to `BackgroundPlotter` either.

The CIs come back happy after this test.

GuillaumeFavelier commented:
I'm interested by your feedback after this test @banesullivan, @akaszynski.

GuillaumeFavelier commented:
Anyone else wants to give a hand or share feedbacks @pyvista/developers ?

akaszynski commented:
Sorry, give me an hour for dinner.

GuillaumeFavelier commented:
> I didn't look at the code in details but the animation looks awesome

I'll check it out now

GuillaumeFavelier commented:
I reverted the test commit, removed the test `bgplot.py`, updated the branch with current `master` and started another test on `mne-python`: https://github.com/mne-tools/mne-python/pull/7241/commits/2cfd5fa90a4418053cd372a392067b585a0ba7a5

GuillaumeFavelier commented:
I obtain: `AssertionError: No output when using VTKs legacy reader` in `volume.py` now.

GuillaumeFavelier commented:
I opened an issue about it: https://github.com/pyvista/pyvista/issues/578

banesullivan commented:
@GuillaumeFavelier, we ran out of bandwidth for git LFS and an example dataset in volume.py is the culprit As GitHub isn’t letting us access that file until next month. We need to move that file elsewhere or remove that particular example

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

banesullivan commented:
It’s the volcano volume 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>
